python/spear/sp_engine.py

```diff
@@ -0,0 +1,271 @@
+#
+# Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+#
+
+import msgpackrpc
+import os
+import psutil
+import spear
+from subprocess import Popen
+import sys
+import time
+
+class SpEngine():
+    def __init__(self, config):
+
+        self._config = config
+        self._rpc_client: msgpackrpc.Client
+
+        self._request_launch_unreal_instance()
+        self._initialize_rpc_client()
+
+        # Need to do these after we have a valid rpc_client
+        self.engine_service = spear.EngineService(self._rpc_client)
+        self.navmesh_service = spear.NavMeshService(self.engine_service)
+
+        # Need to do this after we have a valid engine_service object because we call begin_tick(), tick(), and end_tick() here.
+        self._initialize_unreal_instance()
+
+    def close(self):
+        # Note that in the constructor, we launch the Unreal instance first and then initialize the RPC client. Normally,
+        # we would do things in the reverse order here. But if we close the client first, then we can't send a command to
+        # the Unreal instance to close it. So we close the Unreal instance first and then close the client.
+        self._request_close_unreal_instance()
+        self._close_rpc_client()
+
+    def _request_launch_unreal_instance(self):
+
+        if self._config.SPEAR.LAUNCH_MODE == "running_instance":
+            spear.log('SPEAR.LAUNCH_MODE == "running_instance" so we assume that the Unreal instance has already launched...')
+            return
+
+        spear.log("Launching Unreal instance...")
+
+        # write temp file
+        temp_dir = os.path.realpath(os.path.join(self._config.SPEAR.TEMP_DIR))
+        temp_config_file = os.path.realpath(os.path.join(temp_dir, "config.yaml"))
```

Comment:
More friendly to the case where we are launching multiple instances simultaneously:

```
temp_config_file = os.path.realpath(os.path.join(temp_dir, config.SPEAR.TEMP_CONFIG_FILE))
```